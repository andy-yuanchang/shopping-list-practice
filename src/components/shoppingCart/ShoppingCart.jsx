import PropTypes from 'prop-types';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import * as fromMessageHelper from '../../js/messageHelper';
import {
  addOrder,
} from '../../redux/actions/orderActions';
import {
  addItem, clearAllItems, deleteItem,
} from '../../redux/actions/shoppingCartActions';
import Button from '../button/Button';
import './ShoppingCart.less';

function ShoppingCart(props) {
  const dispatch = useDispatch();
  const itemList = useSelector((state) => state.shoppingCart.list);
  const { onClose } = props;

  const handleAddItem = (item) => {
    dispatch(addItem({
      storeName: item.storeName,
      name: item.name,
      count: 1,
    }));
  };

  const handleDeleteItem = (item) => dispatch(deleteItem({
    storeName: item.storeName,
    name: item.name,
    count: 1,
  }));

  const handleAddOrder = () => {
    if (itemList.length === 0) {
      return fromMessageHelper.addWarning('Your Cart is Empty!');
    }
    dispatch(addOrder({
      itemList,
    }));
    dispatch(clearAllItems());
    fromMessageHelper.addMessageSaved();
    return onClose();
  };

  return (
    <div className="shopping-cart">
      <h1 className="title">
        Items in Cart
      </h1>
      {
        itemList.map((item, index) => (
          <div className="item" key={index}>
            <div className="store-name">{item.storeName}</div>
            <div className="item-name">{item.name}</div>
            <div className="count-area">
              <div
                className="decrease"
                onClick={() => handleDeleteItem(item)}
                role="button"
                tabIndex={0}
                onKeyDown={undefined}
                aria-label="decrease a item"
              />
              <div className="count">{item.count}</div>
              <div
                className="increase"
                onClick={() => handleAddItem(item)}
                role="button"
                tabIndex={0}
                onKeyDown={undefined}
                aria-label="increase a item"
              />
            </div>
          </div>
        ))
      }
      <div className="footer">
        <Button
          text="Order"
          onClick={handleAddOrder}
          size="large"
        />
      </div>
    </div>
  );
}

ShoppingCart.propTypes = {
  onClose: PropTypes.func.isRequired,
};

export default ShoppingCart;
