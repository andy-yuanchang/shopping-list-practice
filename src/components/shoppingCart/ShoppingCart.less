@import '../generalAttribute.less';

.shopping-cart {
  display: flex;
  justify-content: center;
  align-items: center;
  align-content: space-around;
  flex-wrap: wrap;
  .item {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-basis: 60%;
    flex-wrap: wrap;
    border: 1px solid @main-color;
    border-radius: 4px;
    padding: 0.5rem;
    margin: 0.5rem;
    .store-name {
      flex-basis: 100%;
      text-align: center;
    }
    .item-name {
      flex-basis: 100%;
      color: @main-color;
      text-align: center;
    }
    .count-area {
      display: flex;
      justify-content: flex-end;
      align-items: center;
      margin-right: 0.5rem;
      flex-basis: 100%;
      .decrease {
        &::before {
          .minus-sign-style();
        }
      }
      .count {
        width: 2rem;
        margin: 0 0.5rem;
        text-align: center;
      }
      .increase {
        &::before {
          .plus-sign-style();
        }
      }
    }
  }
  .footer {
    flex-basis: 60%;
    display: flex;
    justify-content: center;
    align-items: center;
  }
}