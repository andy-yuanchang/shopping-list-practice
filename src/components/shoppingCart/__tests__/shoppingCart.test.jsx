import React from 'react'
import ReactDOM from 'react-dom'
import { act } from "react-dom/test-utils"
import { render, cleanup } from '@testing-library/react'
import TestRenderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'

import ShoppingCart from '../ShoppingCart'

afterEach(cleanup);
const mockStore = configureMockStore()

function TestShoppingCart({ store }) {
  return (
    <Provider store={store}>
      <ShoppingCart 
        onClose={() => {
          
        }}
      />
    </Provider>
  )
}

describe('OrderedList', () => {
  it('default view', () => {
    const store = mockStore({
      shoppingCart: {
        list: [{
          storeName: 'kebuke',
          name: 'bubble tea',
          count: '5'
        }, {
          storeName: 'i lan',
          name: 'rumen',
          count: '2'
        }, {
          storeName: '7-11',
          name: 'coke',
          count: '15'
        }, {
          storeName: 'family mart',
          name: 'beer',
          count: '24'
        }]
      }
    })
    const shoppingCart = TestRenderer.create(
      <TestShoppingCart store={store} />
    )
    expect(shoppingCart.toJSON()).toMatchSnapshot()
  })

  it('empty view', () => {
    const store = mockStore({ shoppingCart: { list: [] } })
    const shoppingCart = TestRenderer.create(
      <TestShoppingCart store={store} />
    )
    expect(shoppingCart.toJSON()).toMatchSnapshot()
  })
})