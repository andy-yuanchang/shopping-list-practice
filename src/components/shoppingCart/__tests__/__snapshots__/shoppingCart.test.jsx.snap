// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderedList default view 1`] = `
<div
  className="shopping-cart"
>
  <div
    className="item"
  >
    <div
      className="store-name"
    >
      kebuke
    </div>
    <div
      className="item-name"
    >
      bubble tea
    </div>
    <div
      className="count-area"
    >
      <div
        className="decrease"
        onClick={[Function]}
      />
      <div
        className="count"
      >
        5
      </div>
      <div
        className="increase"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="item"
  >
    <div
      className="store-name"
    >
      i lan
    </div>
    <div
      className="item-name"
    >
      rumen
    </div>
    <div
      className="count-area"
    >
      <div
        className="decrease"
        onClick={[Function]}
      />
      <div
        className="count"
      >
        2
      </div>
      <div
        className="increase"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="item"
  >
    <div
      className="store-name"
    >
      7-11
    </div>
    <div
      className="item-name"
    >
      coke
    </div>
    <div
      className="count-area"
    >
      <div
        className="decrease"
        onClick={[Function]}
      />
      <div
        className="count"
      >
        15
      </div>
      <div
        className="increase"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="item"
  >
    <div
      className="store-name"
    >
      family mart
    </div>
    <div
      className="item-name"
    >
      beer
    </div>
    <div
      className="count-area"
    >
      <div
        className="decrease"
        onClick={[Function]}
      />
      <div
        className="count"
      >
        24
      </div>
      <div
        className="increase"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="footer"
  >
    <div
      className="button large"
      onClick={[Function]}
    >
      Order
    </div>
  </div>
</div>
`;

exports[`OrderedList empty view 1`] = `
<div
  className="shopping-cart"
>
  <div
    className="footer"
  >
    <div
      className="button large"
      onClick={[Function]}
    >
      Order
    </div>
  </div>
</div>
`;
